src/irmin-pack/unix/control_file_intf.ml

```diff
@@ -103,7 +103,60 @@ module Payload_v3 = struct
       extensions *)
 end
 
-module Latest_payload = Payload_v3
+module Payload_v4 = struct
+  type from_v4_gced = {
+    suffix_start_offset : int63;
+    generation : int;
+    oldest_live_commit_offset : int63;
+  }
+  [@@deriving irmin]
+  (** Similar to [from_v3_gced]. [oldest_live_commit_offset] is the commit on
+      which the latest gc was called on V4 stores. *)
+
+  (** [From_v1_v2_post_upgrade] similar to [Payload_v3.From_v1_v2_post_upgrade]
+
+      [No_gc_yet] corresponds to a pack store that was created using [`V3] or
+      above, code. It never underwent a GC.
+
+      [Used_non_minimal_indexing_strategy] corresponds to a pack store that was
+      created using [`V3] or above, code. It never underwent a GC and it will
+      never be possible to GC it because entries were pushed using a non-minimal
+      indexing strategy.
+
+      [Gced_3] is a [`V3] store that was GCed at least once.
+
+      [Gced_4] is a [`V4] store that was GCed at least once.
+
+      The [T*] tags are provisional tags that the binary decoder is aware of and
+      that may in the future be used to add features to the [`V3] payload. *)
+  type status =
+    | From_v1_v2_post_upgrade of Payload_v3.from_v1_v2_post_upgrade
+    | No_gc_yet
+    | Used_non_minimal_indexing_strategy
+    | Gced_3 of Payload_v3.from_v3_gced
```

Comment:
The current distinction between `Gced_3` and `Gced_4` provides 2 things:
1. remembering that the store was upgraded from 3 to 4 and
2. not having to bother with the `oldest_live_commit_offset` during upgrade,

---

About `1.` I believe we should remember this for all tags, not just `Gced`. 

Duplicating all tags to remember about upgrade doesn't scale well across upgrades.

I suggest we add a boolean flag in `t` below to remember that (e.g. `upgraded_from_v3_to_v4 : bool;`).

---

About `2.` I believe we should compute the `oldest_live_commit_offset` for the old `V3.From_v3_gced` tags (we know that in `V3` stores this is the start offset of the suffix).

Computing `oldest_live_commit_offset` during upgrade can't be done from within the CF scope, it must be done by the FM. That's why I believe that the `convert_v3_to_v4` function should go and instead be an `upgrade_payload` function implemented within the FM and passed to `Cf.open_`. 

The signature for `upgrade_payload` could be `[ 'V3 of Payload_v3.t ] -> Latest_payload.t`.

Reply:
Added `upgraded_from_v3_to_v4` (I'm not sure why we need it though).

Good point about V3 stores and `oldest_live_commit_offset`. 

About the upgrade function, I did as you suggest, I think. We have to pass `upgrade` to reload too, though.

Reply:
> I'm not sure why we need it though

Me neither, yet.

>  I think. We have to pass upgrade to reload too,

👍

Reply:
The only general scenario I can think of is that the flag would help us distinguish between stores that were upgraded versus those created fresh at v4. Maybe this would help debug an invalid/corrupted store, but I can't immediately think of a concrete scenario to justify carrying the flag around just in case. Is there another implicit marker that already exists for distinguishing this case? Some other reason for the flag? :thinking: (since we are early in the dev process we can always remove it later -- or add it back -- if we change our minds in either direction)

Reply:
> Maybe this would help debug an invalid/corrupted store

Yes, I think that's a big reason to include this

Reply:
Yeah, I guess we don't have enough history to know if this kind of flag would be useful or not in that scenario, so I am on the fence about including it or not. :shrug: